tiledb/sm/array_schema/dimension.cc

```diff
@@ -1700,412 +1558,50 @@ void Dimension::ensure_datatype_is_supported(Datatype type) const {
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      tile_extent_uint64 = (const uint64_t*)tile_extent_.data();
-      ss << *tile_extent_uint64;
-      return ss.str();
-
-    case Datatype::BLOB:
-    case Datatype::CHAR:
-    case Datatype::BOOL:
-    case Datatype::STRING_ASCII:
-    case Datatype::STRING_UTF8:
-    case Datatype::STRING_UTF16:
-    case Datatype::STRING_UTF32:
-    case Datatype::STRING_UCS2:
-    case Datatype::STRING_UCS4:
-    case Datatype::ANY:
+  auto g = [&](auto T) {
+    if constexpr (std::is_same_v<decltype(T), char>) {
       // Not supported domain type
       assert(false);
-      return "";
-  }
-
-  assert(false);
-  return "";
+      return std::string("");
+    }
+    auto val = reinterpret_cast<const decltype(T)*>(tile_extent_.data());
+    ss << *val;
+    return ss.str();
+  };
+  return execute_callback_with_type(type_, g);
 }
 
 void Dimension::set_crop_range_func() {
-  switch (type_) {
-    case Datatype::INT32:
-      crop_range_func_ = crop_range<int32_t>;
-      break;
-    case Datatype::INT64:
-      crop_range_func_ = crop_range<int64_t>;
-      break;
-    case Datatype::INT8:
-      crop_range_func_ = crop_range<int8_t>;
-      break;
-    case Datatype::UINT8:
-      crop_range_func_ = crop_range<uint8_t>;
-      break;
-    case Datatype::INT16:
-      crop_range_func_ = crop_range<int16_t>;
-      break;
-    case Datatype::UINT16:
-      crop_range_func_ = crop_range<uint16_t>;
-      break;
-    case Datatype::UINT32:
-      crop_range_func_ = crop_range<uint32_t>;
-      break;
-    case Datatype::UINT64:
-      crop_range_func_ = crop_range<uint64_t>;
-      break;
-    case Datatype::FLOAT32:
-      crop_range_func_ = crop_range<float>;
-      break;
-    case Datatype::FLOAT64:
-      crop_range_func_ = crop_range<double>;
-      break;
-    case Datatype::DATETIME_YEAR:
-    case Datatype::DATETIME_MONTH:
-    case Datatype::DATETIME_WEEK:
-    case Datatype::DATETIME_DAY:
-    case Datatype::DATETIME_HR:
-    case Datatype::DATETIME_MIN:
-    case Datatype::DATETIME_SEC:
-    case Datatype::DATETIME_MS:
-    case Datatype::DATETIME_US:
-    case Datatype::DATETIME_NS:
-    case Datatype::DATETIME_PS:
-    case Datatype::DATETIME_FS:
-    case Datatype::DATETIME_AS:
-    case Datatype::TIME_HR:
-    case Datatype::TIME_MIN:
-    case Datatype::TIME_SEC:
-    case Datatype::TIME_MS:
-    case Datatype::TIME_US:
-    case Datatype::TIME_NS:
-    case Datatype::TIME_PS:
-    case Datatype::TIME_FS:
-    case Datatype::TIME_AS:
-      crop_range_func_ = crop_range<int64_t>;
-      break;
-    default:
-      crop_range_func_ = nullptr;
-      break;
-  }
+  auto g = [&](auto T) { crop_range_func_ = crop_range<decltype(T)>; };
```

Comment:
Here and a few places below... We don't support char here in the original implementation... It should be filtered out and crop_range_func_ should be set to null.